framework/src/meshgenerators/RenameBlockGenerator.C

```diff
@@ -22,135 +24,272 @@ RenameBlockGenerator::validParams()
+      {
+        block_ids.insert(id);
+        return id;
+      }
+
+    mooseError("Failed to find an unused ID!");
+  };
+
+  const auto num_blocks = _old_block.size();
+
+  // Helper for checking whether or not a SubdomainName (which could be an ID or a name)
+  // is really input as an ID
+  const auto is_subdomain_id = [](const SubdomainName & subdomain_name) {
+    SubdomainID id = Moose::INVALID_BLOCK_ID;
+    std::istringstream ss(subdomain_name);
+    return !(!(ss >> id));
+  };
+
+  // Get the old block IDs and make sure they exist
+  std::vector<SubdomainID> old_block_ids(num_blocks, Moose::INVALID_BLOCK_ID);
+  std::vector<SubdomainName> old_block_names(num_blocks);
+  std::stringstream missing_block;
+  for (const auto i : make_range(num_blocks))
   {
-    // user must have supplied old_block_name
-    _old_block_name = getParam<std::vector<SubdomainName>>("old_block_name");
-    _old_block_id.reserve(_old_block_name.size());
-    for (const auto & block_name : _old_block_name)
-      _old_block_id.emplace_back(mesh->get_id_by_name(block_name));
+    const SubdomainName & name = _old_block[i];
+
+    // Convert the SubdomainName to an id and store
+    const auto id = MooseMeshUtils::getSubdomainID(name, *mesh);
+    old_block_ids[i] = id;
+
+    // Block does not exist - store for a future error
+    if (!block_ids.count(id))
+      missing_block << name << " ";
+
+    // Keep track of the block names
+    // If this SubdomainName is an ID, try to see if it has a name set
+    if (is_subdomain_id(name))
+      old_block_names[i] = const_mesh.subdomain_name(id);
+    // If this SubdomainName is a name, use said name
+    else
+      old_block_names[i] = name;
   }
+  if (missing_block.str().size())
+    paramError(_old_block_param_name,
+               "The following blocks were requested to be renamed, but do not exist: ",
+               missing_block.str());
 
-  // construct new_block_id and new_block_name
-  if (isParamValid("new_block_id"))
+  // Get the block IDs that we're moving to
+  std::vector<SubdomainID> new_block_ids(num_blocks, Moose::INVALID_BLOCK_ID);
+  std::map<SubdomainID, std::string> new_names;
+  for (const auto i : make_range(num_blocks))
   {
-    _new_block_id = getParam<std::vector<subdomain_id_type>>("new_block_id");
-    if (_new_block_id.size() != _old_block_id.size())
-      mooseError("RenameBlockGenerator: The vector of old_block information must have the same"
-                 " length as the vector of new_block information\n");
-
-    // construct the _new_block_name
-    _new_block_name.reserve(_new_block_id.size());
-    for (const auto & block_id : _new_block_id)
-      _new_block_name.emplace_back(newBlockName(block_id));
+    const SubdomainName & name = _new_block[i];
+
+    // If the user input an ID, we have the ID
+    if (is_subdomain_id(name))
+    {
+      const auto id = MooseMeshUtils::getSubdomainID(name, *mesh);
```

Comment:
no, it's to check whether it's an ID or a name

Reply:
but this does it 
(is_subdomain_id(name)) ? not this

      const auto id = MooseMeshUtils::getSubdomainID(name, *mesh);

Reply:
No. All that does is convert a name to an ID if it it's a name. It doesn't tell you if what you actually passed in is an ID or not

Reply:
You said:
no, it's to check whether it's an ID or a name
for
      const auto id = MooseMeshUtils::getSubdomainID(name, *mesh);

I said:
    if (is_subdomain_id(name))
does " check whether it's an ID or a name "
and 
      const auto id = MooseMeshUtils::getSubdomainID(name, *mesh);
doesnt do much since we already know it's an id
eg there is no need to convert ?

Reply:
Let's start over.

`getSubdomainID(subdomain_name)` converts a string that is an ID to a block ID, or a string that is a subdomain name to the ID associated with that name (if it exists in the subdomain map).

This additional `is_subdomain_id` is needed to check whether or not a name is explicitly an ID vs a name, because the logic that follows depends on whether or not it was actually provided as an ID or as a name that we get the ID from.